docs/protocol.md

```diff
@@ -0,0 +1,364 @@
+# HTTP protocol for RAG chat apps
+
+We are standardizing on a common HTTP protocol across RAG chat app solutions and tools,
+to make them more compatible with each other. By agreeing on a common protocol, developers
+can swap different components, like using the Python backend from this repository
+with a Web Components frontend from [azure-search-openai-javascript](https://github.com/Azure-Samples/azure-search-openai-javascript).
+
+This protocol is inspired by the [OpenAI ChatCompletion API](https://platform.openai.com/docs/guides/text-generation/chat-completions-api),
+but contains additional fields required for a RAG chat app.
+
+## HTTP requests to RAG chat app endpoints
+
+An HTTP request to `/chat` or `/ask` can contain these properties, in JSON:
+
+* `"messages"`: A list of messages, each containing "content" and "role", where "role" may be "system", "assistant", or "user"
+* `"stream"`: A boolean indicating whether the response should be streamed or not.
+* `"context"`: An object containing any additional context about the request. This app only sets the "overrides" property, which is an object with additional settings for the request, like how many search results to return.
+* `"session_state"`: An object containing the "memory" for the chat app, such as a user ID. Not currently used by this app.
+
+Here's an example JSON request:
+
+```json
```

Comment:
We are missing a version field in the protocol, we might want to add / remove fields in the future. How will clients for version X deal with version X+1, X-1 from the server?

Reply:
I think the version field is helpful for conveying that things changed, but also practically, we're not intending to serve older versions at same time, so I think either a client supports version X+1 or it doesn't, right?
It's good to have though, for communication purposes.

Reply:
Version updates will break clients, IMO if there's no clear version field there will be issues with mismatched clients / servers

Reply:
I've added a version number to top of the doc in new repo-
https://github.com/Azure-Samples/ai-chat-app-protocol/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L1

And filed an issue with more thoughts:
https://github.com/Azure-Samples/ai-chat-app-protocol/issues/2